### ISSUE TITLE
Add 410 Gone to http-request deny codes

### ISSUE THREAD

**Issue by TechyShishy** [haproxy/haproxy#80]
## Output of `haproxy -vv` and `uname -a`
```
whirsch@lb-1:~$ haproxy -vv
HA-Proxy version 1.8.14-52e4d43 2018/09/20
Copyright 2000-2018 Willy Tarreau <willy@haproxy.org>

Build options :
  TARGET  = linux2628
  CPU     = x86_64
  CC      = gcc
  CFLAGS  = -g -fno-strict-aliasing -Wdeclaration-after-statement -fwrapv -fno-strict-overflow -Wno-unused-label
  OPTIONS = USE_LINUX_SPLICE=1 USE_LINUX_TPROXY=1 USE_LIBCRYPT=1 USE_ZLIB=1 USE_OPENSSL=1 USE_PCRE=1

Default settings :
  maxconn = 2000, bufsize = 16384, maxrewrite = 1024, maxpollevents = 200

Built with OpenSSL version : OpenSSL 1.0.1f 6 Jan 2014
Running on OpenSSL version : OpenSSL 1.0.1f 6 Jan 2014
OpenSSL library supports TLS extensions : yes
OpenSSL library supports SNI : yes
OpenSSL library supports : SSLv3 TLSv1.0 TLSv1.1 TLSv1.2
Built with transparent proxy support using: IP_TRANSPARENT IPV6_TRANSPARENT IP_FREEBIND
Encrypted password support via crypt(3): yes
Built with multi-threading support.
Built with PCRE version : 8.31 2012-07-06
Running on PCRE version : 8.31 2012-07-06
PCRE library supports JIT : no (USE_PCRE_JIT not set)
Built with zlib version : 1.2.8
Running on zlib version : 1.2.8
Compression algorithms supported : identity("identity"), deflate("deflate"), raw-deflate("deflate"), gzip("gzip")
Built with network namespace support.

Available polling systems :
      epoll : pref=300,  test result OK
       poll : pref=200,  test result OK
     select : pref=150,  test result OK
Total: 3 (3 usable), will use epoll.

Available filters :
	[SPOE] spoe
	[COMP] compression
	[TRACE] trace

whirsch@lb-1:~$ uname -a
Linux lb-1.redacted.tld 3.13.0-123-generic #172-Ubuntu SMP Mon Jun 26 18:04:35 UTC 2017 x86_64 x86_64 x86_64 GNU/Linux

```

## What should haproxy do differently? Which functionality do you think we should add?

Haproxy should support the response of `410 Gone` from `http-request deny`, so as to inform crawlers that page they've requested has been intentionally permanently removed. 

## What are you trying to do?

In an effort to reduce load on my application, I'd like haproxy to respond to bot requests for certain long-dead paths with `410 Gone`, so that the crawler removes them from its corpus, and doesn't attempt to crawl them anymore.

**Comment by wtarreau** [comment]
Normally you can already do it, please have a look at the "deny_status" option of "http-request deny".

**Comment by TimWolla** [comment]
> Normally you can already do it, please have a look at the "deny_status" option of "http-request deny".

This does not work, because `http_err_codes` does not contain 410 (and neither 404).

**Comment by wtarreau** [comment]
On Mon, Apr 22, 2019 at 04:30:44AM -0700, Tim Düsterhus wrote:
> > Normally you can already do it, please have a look at the "deny_status" option of "http-request deny".
> 
> This does not work, because `http_err_codes` does not contain 410 (and neither 404).

Ah indeed, I thought we could emit any code but that's not the case. Thus
we should probably add a new code here.

Willy

**Comment by TechyShishy** [comment]
404 would be nice too as an available status code.

Is there a technical reason that haproxy _shouldn't_ be able to emit any code specified?

**Comment by wtarreau** [comment]
On Mon, Apr 22, 2019 at 06:50:02AM -0700, Wesley Hirsch wrote:
> 404 would be nice too as an available status code.
> 
> Is there a technical reason that haproxy _shouldn't_ be able to emit any code
> specified?

We do have pre-built response messages which are sent as-is on the wire,
and which can be replaced using errorfiles. I *thought* it was possible
to send any of them by now, which is not the case. The main limitation
probably is only the reason string and the ability, which is useless and
was dropped in H2 (and we already have a function to emit it anyway), and
the ability to load errorfiles for arbitrary codes.

This is probably something which should be revisited at some point in time.

Willy

**Comment by capflam** [comment]
Now supported. Thanks !

**Comment by cherouvim** [comment]
I run 2.1 on Ubuntu 20 via `ppa:vbernat/haproxy-2.1`:
```
$ haproxy -vv
HA-Proxy version 2.1.7-1ppa1~focal 2020/06/12 - https://haproxy.org/
Status: stable branch - will stop receiving fixes around Q1 2021.
Known bugs: http://www.haproxy.org/bugs/bugs-2.1.7.html
Running on: Linux 5.4.0-26-generic #30-Ubuntu SMP Mon Apr 20 16:58:30 UTC 2020 x86_64
Build options :
  TARGET  = linux-glibc
  CPU     = generic
  CC      = gcc
  CFLAGS  = -O2 -g -O2 -fdebug-prefix-map=/build/haproxy-Gr2xsC/haproxy-2.1.7=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -fno-strict-aliasing -Wdeclaration-after-statement -fwrapv -Wno-address-of-packed-member -Wno-unused-label -Wno-sign-compare -Wno-unused-parameter -Wno-old-style-declaration -Wno-ignored-qualifiers -Wno-clobbered -Wno-missing-field-initializers -Wno-implicit-fallthrough -Wno-stringop-overflow -Wno-cast-function-type -Wtype-limits -Wshift-negative-value -Wshift-overflow=2 -Wduplicated-cond -Wnull-dereference
  OPTIONS = USE_PCRE2=1 USE_PCRE2_JIT=1 USE_REGPARM=1 USE_OPENSSL=1 USE_LUA=1 USE_ZLIB=1 USE_SYSTEMD=1

Feature list : +EPOLL -KQUEUE -MY_EPOLL -MY_SPLICE +NETFILTER -PCRE -PCRE_JIT +PCRE2 +PCRE2_JIT +POLL -PRIVATE_CACHE +THREAD -PTHREAD_PSHARED +REGPARM -STATIC_PCRE -STATIC_PCRE2 +TPROXY +LINUX_TPROXY +LINUX_SPLICE +LIBCRYPT +CRYPT_H -VSYSCALL +BACKTRACE +GETADDRINFO +OPENSSL +LUA +FUTEX +ACCEPT4 -MY_ACCEPT4 +ZLIB -SLZ +CPU_AFFINITY +TFO +NS +DL +RT -DEVICEATLAS -51DEGREES -WURFL +SYSTEMD -OBSOLETE_LINKER +PRCTL +THREAD_DUMP -EVPORTS

Default settings :
  bufsize = 16384, maxrewrite = 1024, maxpollevents = 200

Built with multi-threading support (MAX_THREADS=64, default=1).
Built with OpenSSL version : OpenSSL 1.1.1f  31 Mar 2020
Running on OpenSSL version : OpenSSL 1.1.1f  31 Mar 2020
OpenSSL library supports TLS extensions : yes
OpenSSL library supports SNI : yes
OpenSSL library supports : TLSv1.0 TLSv1.1 TLSv1.2 TLSv1.3
Built with Lua version : Lua 5.3.3
Built with network namespace support.
Built with transparent proxy support using: IP_TRANSPARENT IPV6_TRANSPARENT IP_FREEBIND
Built with PCRE2 version : 10.34 2019-11-21
PCRE2 library supports JIT : yes
Encrypted password support via crypt(3): yes
Built with zlib version : 1.2.11
Running on zlib version : 1.2.11
Compression algorithms supported : identity("identity"), deflate("deflate"), raw-deflate("deflate"), gzip("gzip")
Built with the Prometheus exporter as a service

Available polling systems :
      epoll : pref=300,  test result OK
       poll : pref=200,  test result OK
     select : pref=150,  test result OK
Total: 3 (3 usable), will use epoll.

Available multiplexer protocols :
(protocols marked as <default> cannot be specified using 'proto' keyword)
              h2 : mode=HTTP       side=FE|BE     mux=H2
            fcgi : mode=HTTP       side=BE        mux=FCGI
       <default> : mode=HTTP       side=FE|BE     mux=H1
       <default> : mode=TCP        side=FE|BE     mux=PASS

Available services :
        prometheus-exporter

Available filters :
        [SPOE] spoe
        [CACHE] cache
        [FCGI] fcgi-app
        [TRACE] trace
        [COMP] compression
```

but the following config still produces a `403` and not a `410`:
```
http-request deny deny_status 410 if { path_reg ^/xxx$ }
```

Am I doing something wrong, or hasn't the fix made it yet to the ppa where from I installed HAProxy?

**Comment by jmagnin** [comment]
@cherouvim this will be available in haproxy 2.2.
